demo_nodes_cpp/src/topics/listener_content_filter.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_components/register_node_macro.hpp"
+
+#include "std_msgs/msg/string.hpp"
+
+#include "demo_nodes_cpp/visibility_control.h"
+
+namespace demo_nodes_cpp
+{
+// Create a Listener class that subclasses the generic rclcpp::Node base class.
+// The main function below will instantiate the class as a ROS node.
+class ListenerContentFilter : public rclcpp::Node
+{
+public:
+  DEMO_NODES_CPP_PUBLIC
+  explicit ListenerContentFilter(const rclcpp::NodeOptions & options)
+  : Node("content_filter_listener", options)
+  {
+    setvbuf(stdout, NULL, _IONBF, BUFSIZ);
+    // Create a callback function for when messages are received.
+    auto callback =
+      [this](const std_msgs::msg::String & msg) -> void
+      {
+        RCLCPP_INFO(this->get_logger(), "I heard: [%s]", msg.data.c_str());
+      };
+    // Create a subscription to the topic which can be matched with one or more compatible ROS
+    // publishers.
+    // Note that not all publishers on the same topic with the same type will be compatible:
+    // they must have compatible Quality of Service policies.
+
+    // Initialize a subscription with a content filter to receive messages,
+    // the default content filter is "data = 'Hello World: 10'".
+    rclcpp::SubscriptionOptions sub_options;
+    sub_options.content_filter_options.filter_expression = "data = %0";
+    std::string data_number = (options.arguments().size() == 2) ? options.arguments()[1] : "10";
```

Comment:
I'm not 100% sure how `.arguments()` works off the top of my head, but couldn't contain more options potentially? It looks fragile to assume that if the size is 2 then it's the arguments you expect...

There's also no usage docs or anything.

Maybe you could just use a ROS parameter instead? Or just not make this configurable from the cli?

Reply:
> Just curious why not put this kind of thing in `ros2/examples` instead?

If put it in `ros2/examples`, I need to add an extra `talker_content_filter.cpp`, which seems unnecessary.

> Maybe you could just use a ROS parameter instead? Or just not make this configurable from the cli?

Considering it's located in `demo_nodes_cpp/src/topics`, and the concept of `ROS parameter` is only used in `demo_nodes_cpp/src/parameters`, I'll take second suggestion to remove the `options.arguments`.